apps/andi/lib/andi/input_schemas/ingestion.ex

```diff
@@ -58,32 +58,75 @@ defmodule Andi.InputSchemas.Ingestion do
     source_format = Map.get(changes, :sourceFormat, nil)
 
     ingestion
-    |> cast(changes_with_id, @cast_fields, empty_values: [])
-    |> validate_required(@required_fields, message: "is required")
-    |> cast_assoc(:schema, with: &DataDictionary.changeset(&1, &2, source_format), invalid_message: "is required")
-    |> cast_assoc(:extractSteps, with: &ExtractStep.changeset/2)
-    |> cast_assoc(:transformations, with: &Transformation.changeset/2)
-    |> foreign_key_constraint(:targetDataset)
+    |> Changeset.cast(changes_with_id, @cast_fields, empty_values: [""])
+    |> Changeset.validate_required(@required_fields, message: "is required")
+    |> Changeset.cast_assoc(:schema, with: &DataDictionary.changeset(&1, &2, source_format), invalid_message: "is required")
+    |> Changeset.cast_assoc(:extractSteps, with: &ExtractStep.changeset/2)
+    |> Changeset.cast_assoc(:transformations, with: &Transformation.changeset/2)
+    |> Changeset.foreign_key_constraint(:targetDataset)
     |> validate_source_format()
     |> CadenceValidator.validate()
     |> validate_top_level_selector()
     |> validate_schema()
     |> validate_extract_steps()
   end
 
-  def changeset_for_draft(ingestion, changes) do
+  def changeset(%Ecto.Changeset{ data: %__MODULE__{} } = changeset, changes) do
+    source_format = Map.get(changes, :sourceFormat, nil)
+
+    changeset
+    |> Changeset.cast(changes, @cast_fields, empty_values: [""])
+    |> Changeset.validate_required(@required_fields, message: "is required")
+    |> Changeset.cast_assoc(:schema, with: &DataDictionary.changeset(&1, &2, source_format), invalid_message: "is required")
+    |> Changeset.cast_assoc(:extractSteps, with: &ExtractStep.changeset/2)
+    |> Changeset.cast_assoc(:transformations, with: &Transformation.changeset/2)
+    |> Changeset.foreign_key_constraint(:targetDataset)
+    |> validate_source_format()
+    |> CadenceValidator.validate()
+    |> validate_top_level_selector()
+    |> validate_schema()
+    |> validate_extract_steps()
+  end
+
+  def changeset_for_draft(%Andi.InputSchemas.Ingestion{} = ingestion, changes) do
     changes_with_id = StructTools.ensure_id(ingestion, changes)
 
     ingestion
-    |> cast(changes_with_id, @cast_fields, empty_values: [])
-    |> cast_assoc(:schema, with: &DataDictionary.changeset_for_draft_ingestion/2)
-    |> cast_assoc(:extractSteps, with: &ExtractStep.changeset_for_draft/2)
-    |> cast_assoc(:transformations, with: &Transformation.changeset_for_draft/2)
+    |> Changeset.cast(changes_with_id, @cast_fields, empty_values: [""])
+    |> Changeset.cast_assoc(:schema, with: &DataDictionary.changeset_for_draft_ingestion/2)
+    |> Changeset.cast_assoc(:extractSteps, with: &ExtractStep.changeset_for_draft/2)
+    |> Changeset.cast_assoc(:transformations, with: &Transformation.changeset_for_draft/2)
+  end
+
+  def changeset_for_draft(%Ecto.Changeset{ data: %__MODULE__{} } = changeset, changes) do
```

Comment:
Same concept as above, except drafts require less validation. 

Note on the draft concept: Ideally, the UI would always use the full ingestion validation and then reject database saves only when publishing. I believe the current pattern allows for invalidated publishes if handled slightly wrong. I didn't look too deeply into this though.